src/github.com/matrix-org/dendrite/appservice/storage/appservice_events_table.go

```diff
@@ -0,0 +1,256 @@
+	deleteEventsBeforeAndIncludingIDStmt   *sql.Stmt
+}
+
+func (s *eventsStatements) prepare(db *sql.DB) (err error) {
+	_, err = db.Exec(appserviceEventsSchema)
+	if err != nil {
+		return
+	}
+
+	if s.selectEventsByApplicationServiceIDStmt, err = db.Prepare(selectEventsByApplicationServiceIDSQL); err != nil {
+		return
+	}
+	if s.countEventsByApplicationServiceIDStmt, err = db.Prepare(countEventsByApplicationServiceIDSQL); err != nil {
+		return
+	}
+	if s.insertEventStmt, err = db.Prepare(insertEventSQL); err != nil {
+		return
+	}
+	if s.updateTxnIDForEventsStmt, err = db.Prepare(updateTxnIDForEventsSQL); err != nil {
+		return
+	}
+	if s.deleteEventsBeforeAndIncludingIDStmt, err = db.Prepare(deleteEventsBeforeAndIncludingIDSQL); err != nil {
+		return
+	}
+
+	return
+}
+
+// selectEventsByApplicationServiceID takes in an application service ID and
+// returns a slice of events that need to be sent to that application service,
+// as well as an int later used to remove these same events from the database
+// once successfully sent to an application service.
+func (s *eventsStatements) selectEventsByApplicationServiceID(
+	ctx context.Context,
+	applicationServiceID string,
+	limit int,
+) (
+	txnID, maxID int,
+	events []gomatrixserverlib.ApplicationServiceEvent,
+	err error,
+) {
+	// Retrieve events from the database. Unsuccessfully sent events first
+	eventRowsCurr, err := s.selectEventsByApplicationServiceIDStmt.QueryContext(ctx, applicationServiceID, limit)
+	if err != nil {
+		return 0, 0, nil, err
+	}
+	defer func() {
+		err = eventRowsCurr.Close()
+		if err != nil {
+			log.WithError(err).Fatalf("Appservice %s unable to select new events to send",
+				applicationServiceID)
+		}
+	}()
+	events, maxID, txnID, err = retrieveEvents(eventRowsCurr)
+	if err != nil {
+		return 0, 0, nil, err
+	}
+
+	return
+}
+
+func retrieveEvents(eventRows *sql.Rows) (events []gomatrixserverlib.ApplicationServiceEvent, maxID, txnID int, err error) {
+	// Get current time for use in calculating event age
+	nowMilli := time.Now().UnixNano() / int64(time.Millisecond)
+
+	// Iterate through each row and store event contents
+	// If txn_id changes dramatically, we've switched from collecting old events to
+	// new ones. Send back those events first.
+	lastTxnID := -2 // Invalid transaction ID
+	for eventRows.Next() {
+		var event gomatrixserverlib.ApplicationServiceEvent
+		var eventContent sql.NullString
+		var id int
+		err = eventRows.Scan(
+			&id,
+			&event.EventID,
+			&event.OriginServerTimestamp,
+			&event.RoomID,
+			&event.Type,
+			&event.UserID,
+			&eventContent,
+			&txnID,
+		)
+		if err != nil {
+			fmt.Println("Failed:", err.Error())
+			return nil, 0, 0, err
+		}
+
+		// If txnID has changed on this event from the previous event, then we've
+		// reached the end of a transaction's events. Return only those events.
+		if lastTxnID > -2 && lastTxnID != txnID {
+			return events, maxID, lastTxnID, nil
```

Comment:
Instead of having this complicated logic for dealing with events that have already been assigned a transaction ID, I'd probably do it as two queries: the first would pull out all events with an assigned txn ID (without a limit), and then if there aren't any events then we can pull out a limited set of events that don't have a txn ID.

This is slightly less efficient but a single DB RTT is probably not going to cause an issue, and if it turns out it does there are a couple of nice ways of skipping the first step in most cases.

Reply:
I have spent quite a few hours trying to reduce this into a single-query ordeal, including talking it over with @krombel. I agree a dual-query setup would be easier to understand, but do you think it is too crazy that we need to switch it back?

Reply:
My main concern is the fact that includes the shotgun of: https://github.com/matrix-org/dendrite/pull/477/files#diff-d84306218ea8217c1024f21eb34d9fcbR39, which feels a little scary

Reply:
Well, we could actually get rid of that if we remove the `LIMIT` in the query, instead only enforcing it in the `for` loop and opting not to enforce it once we've completely sent the last transaction.

Reply:
I have done the about in 801b58c9277511b1b5c26a908dd101f8026db555. Let me know what you think.